compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp

```diff
@@ -747,9 +747,24 @@ std::optional<SmallVector<int64_t>> getMmaNativeVectorSize(Operation *op) {
         // OptimizeVectorTransfer, matrixC reads are moved above the mainloop
         // and writes are moved below the mainloop. Thus, mma.sync read/write
         // accumulator inplace.
-
         SmallVector<int64_t> readShape;
-        readShape.append({16, 16});
+        int outer, inner;
+        // If the read is already transposed we can unroll to the full tiled
+        // block of 16x16. Otherwise we need to match the unrolling of the
+        // contract to ensure conversion to mma.sync won't be blocked by
+        // extract slice ops.
```

Comment:
I left a note in the message above but reiterating here. The pass pipeline based codegen flow for mma.sync will produce transposed vector.transfer_reads when vectorizing and as a result the transpose introduced when converting to nvgpu will be unable to fold into the transfer_reads due to an extract_slice in the middle (https://github.com/llvm/llvm-project/blob/c4039c10282f6cc88650f99e05fc1c382605463a/mlir/lib/Conversion/VectorToGPU/VectorToGPU.cpp#L443)

The solution employed here is to match the unrolling of the contract, which will always be correct although might not be exactly what's wanted for performance (we end up with extra ld_matrix calls by unrolling more).